winpr/include/winpr/collections.h

```diff
@@ -302,66 +289,67 @@ extern "C"
 
 	/* Hash Table */
 
-	typedef UINT32 (*HASH_TABLE_HASH_FN)(void* key);
-	typedef BOOL (*HASH_TABLE_KEY_COMPARE_FN)(void* key1, void* key2);
-	typedef BOOL (*HASH_TABLE_VALUE_COMPARE_FN)(void* value1, void* value2);
-	typedef void* (*HASH_TABLE_KEY_CLONE_FN)(void* key);
-	typedef void* (*HASH_TABLE_VALUE_CLONE_FN)(void* value);
-	typedef void (*HASH_TABLE_KEY_FREE_FN)(void* key);
-	typedef void (*HASH_TABLE_VALUE_FREE_FN)(void* value);
+	typedef UINT32 (*HASH_TABLE_HASH_FN)(const void* key);
 
-	/* WARNING: Do not access structs directly, the API will be reworked
-	 * to make this opaque. */
-	struct _wHashTable
-	{
-		BOOL synchronized;
-		CRITICAL_SECTION lock;
-
-		int numOfBuckets;
-		int numOfElements;
-		float idealRatio;
-		float lowerRehashThreshold;
-		float upperRehashThreshold;
-		wKeyValuePair** bucketArray;
-
-		HASH_TABLE_HASH_FN hash;
-		HASH_TABLE_KEY_COMPARE_FN keyCompare;
-		HASH_TABLE_VALUE_COMPARE_FN valueCompare;
-		HASH_TABLE_KEY_CLONE_FN keyClone;
-		HASH_TABLE_VALUE_CLONE_FN valueClone;
-		HASH_TABLE_KEY_FREE_FN keyFree;
-		HASH_TABLE_VALUE_FREE_FN valueFree;
-
-		DWORD foreachRecursionLevel;
-		DWORD pendingRemoves;
-	};
 	typedef struct _wHashTable wHashTable;
 
-	typedef BOOL (*HASH_TABLE_FOREACH_FN)(const void* key, void* value, void* arg);
+	typedef BOOL (*HASH_TABLE_FOREACH_FN)(const void* key, const void* value, void* arg);
```

Comment:
IMHO setting value as `const` is a bad idea, as in many cases we will want to modify `value`, and so we will have to break the const cast.

Reply:
You´re right, removed.